README.md
# Stripy [![hex.pm](https://img.shields.io/hexpm/v/stripy.svg?style=flat-square)](https://hex.pm/packages/stripy) [![hexdocs.pm](https://img.shields.io/badge/docs-latest-green.svg?style=flat-square)](https://hexdocs.pm/stripy)

Stripy is a micro wrapper intended to be
used for sending requests to Stripe's REST API. It is
made for developers who prefer to work directly with the
official API and provide their own abstractions on top
if such are needed.

Stripy takes care of setting headers, encoding the data,
configuration settings, etc (the usual boring boilerplate);
it also makes testing easy by letting you plug your own
mock server (see Testing section below).

Some basic examples:

```elixir
iex> Stripy.req(:get, "subscriptions")
{:ok, %HTTPoison.Response{...}}

iex> Stripy.req(:post, "customers", %{"email" => "a@b.c", "metadata[user_id]" => 1})
{:ok, %HTTPoison.Response{...}}
```

Where `subscriptions` and `customers` are [REST API resources](https://stripe.com/docs/api).

If you prefer to work with a higher-level library, check out
"stripity_stripe" or "stripe_elixir" on Hex.

## Installation

Add to your `mix.exs` as usual:
```elixir
def deps do
  [{:stripy, "~> 2.0"}]
end
```
If you're not using [application inference](https://elixir-lang.org/blog/2017/01/05/elixir-v1-4-0-released/#application-inference), then add `:stripy` to your `applications` list.

Then configure the `stripy` app per environment like so:

```elixir
config :stripy,
  secret_key: "sk_test_xxxxxxxxxxxxx", # required
  endpoint: "https://api.stripe.com/v1/", # optional
  version: "2017-06-05", # optional
  httpoison: [recv_timeout: 5000, timeout: 8000] # optional
```

You may also use environment variables:

``` elixir
config :stripy,
  secret_key: {:system, "STRIPE_SECRET_KEY"},
  endpoint: {:system, "STRIPE_ENDPOINT"},
  version: {:system, "STRIPE_VERSION"}
```

## Testing

You can disable actual calls to the Stripe API like so:

```elixir
# Usually in your test.exs.
config :stripy,
  testing: true
```

All functions that use Stripy would receive response `{:ok, %{status_code: 200, body: "{}"}}`.

To provide your own responses, you need to configure a mock server:

```elixir
config :stripy,
  testing: true,
  mock_server: MyApp.StripeMockServer
```

Here's an example mock server that mocks the `/customer` endpoint and returns a basic
object for a customer with id `cus_test`

```elixir
defmodule MyApp.StripeMockServer do
  @behaviour Stripy.MockServer

  @ok_res %{status_code: 200}

  @impl Stripy.MockServer
  def request(:get, "customers/cus_test", %{}) do
    body = Poison.encode!(%{"email" => "email@email.com"})
    {:ok, Map.put(@ok_res, :body, body)}
  end
end
```

Now let's quickly write a naive function that gets user's billing email:

```elixir
def stripe_email(user) do
  {:ok, res} = Stripy.req(:get, "customers/#{user.stripe_id}")
  res["email"]
end
```

We can test it like so:

```elixir
fake_user = %{stripe_id: "cus_test"}
assert stripe_email(fake_user) == "email@email.com"
```

## Custom headers

You can add custom headers to the request by supplying a fourth parameter:

```elixir
Stripy.req(:post, "charges", %{amount: 1000}, %{"Idempotency-Key" => "123456"})
```

## License

- Stripy: See LICENSE file.


CHANGELOG.md
# Changelog

# 2.1.0 (2020-02-11)
## New features
- Allow fetching config at runtime, for example `{:system, "ENV_VAR"}` in `config.exs`

# 2.0.0 (2020-02-11)
## Breaking changes
- Removed `headers/1` in favour of new `req/4` (see below)
- Bumped required version of Elixir to `~> 1.7`

## New features
- `req/4` accepts a fourth optional parameter that allows you to send extra
headers to Stripe, e.g.

```elixir
%{"Stripe-Account" => "...", "Idempotency-Key" => "..."}
```

# 1.2.1 (2018-12-05)
- Relaxed `HTTPoison` dependency version.

# 1.2.0 (2018-03-16)
- Introducing `testing` mode and `Stripy.MockServer` behaviour for easier testing.

# 1.1.0 (2018-03-15)
- Ability to pass options to HTTPoison through the `httpoison` config.

# 1.0.0
- First version


test/stripy_test.exs
defmodule StripyTest do
  use ExUnit.Case
end


mix.exs
defmodule Stripy.Mixfile do
  use Mix.Project

  @version "2.1.0"

  def project do
    [
      app: :stripy,
      version: @version,
      elixir: "~> 1.7",
      name: "Stripy",
      description: "Micro wrapper for the Stripe REST API",
      package: package(),
      docs: [
        main: "readme",
        extras: ["README.md"],
        source_url: "https://github.com/svileng/stripy",
        source_ref: @version
      ],
      deps: deps()
    ]
  end

  def application do
    [extra_applications: [:logger]]
  end

  defp deps do
    [
      {:httpoison, "~> 0.13 or ~> 1.0"},
      {:poison, "~> 3.1"},
      {:ex_doc, "~> 0.21", only: :dev, runtime: false}
    ]
  end

  defp package do
    [
      maintainers: ["Svilen Gospodinov <webmaster@s2g.io>"],
      licenses: ["MIT"],
      links: %{GitHub: "https://github.com/svileng/stripy"}
    ]
  end
end


lib/stripy.ex
defmodule Stripy do
  @moduledoc """
  Stripy is a micro wrapper intended to be
  used for sending requests to Stripe's REST API. It is
  made for developers who prefer to work directly with the
  official API and provide their own abstractions on top
  if such are needed.

  Stripy takes care of setting headers, encoding the data,
  configuration settings, etc (the usual boring boilerplate);
  it also provides a `parse/1` helper function for decoding.

  Some basic examples:

      iex> Stripy.req(:get, "subscriptions")
      {:ok, %HTTPoison.Response{...}}

      iex> Stripy.req(:post, "customers", %{"email" => "a@b.c", "metadata[user_id]" => 1})
      {:ok, %HTTPoison.Response{...}}

  You are expected to build your business logic on top
  of Stripy and abstract things such as Subscriptions
  and Customers; if that's not your cup of tea,
  check out "stripity_stripe" or "stripe_elixir" on Hex.
  """

  @content_type_header %{"Content-Type" => "application/x-www-form-urlencoded"}

  @doc "Constructs url with query params from given data."
  def url(api_url, resource, data) do
    api_url <> resource <> "?" <> URI.encode_query(data)
  end

  @doc """
  Makes request to the Stripe API.

  Will return an HTTPoison standard response; see `parse/1`
  for decoding the response body.

  You can specify custom headers to be included in the request
  to Stripe, such as `Idempotency-Key`, `Stripe-Account` or any
  other header. Just pass a map as the fourth argument.
  See example below.

  ## Examples
      iex> Stripy.req(:get, "subscriptions")
      {:ok, %HTTPoison.Response{...}}

      iex> Stripy.req(:post, "customers", %{"email" => "a@b.c", "metadata[user_id]" => 1})
      {:ok, %HTTPoison.Response{...}}

      iex> Stripy.req(:post, "customers", %{"email" => "a@b.c"}, %{"Idempotency-Key" => "ABC"})
      {:ok, %HTTPoison.Response{...}}
  """
  def req(action, resource, data \\ %{}, headers \\ %{}) when action in [:get, :post, :delete] do
    if Application.get_env(:stripy, :testing, false) do
      mock_server = Application.get_env(:stripy, :mock_server, Stripy.MockServer)
      mock_server.request(action, resource, data)
    else
      secret_key = get_config!(:secret_key)

      headers =
        @content_type_header
        |> Map.merge(%{
          "Authorization" => "Bearer #{secret_key}",
          "Stripe-Version" => get_config(:version, "2017-06-05")
        })
        |> Map.merge(headers)
        |> Map.to_list()

      api_url = get_config(:endpoint, "https://api.stripe.com/v1/")

      options = Application.get_env(:stripy, :httpoison, [])

      url = url(api_url, resource, data)
      HTTPoison.request(action, url, "", headers, options)
    end
  end

  @doc "Parses an HTTPoison response from a Stripe API call."
  def parse({:ok, %{status_code: 200, body: body}}) do
    {:ok, Poison.decode!(body)}
  end

  def parse({:ok, %{body: body}}) do
    error = Poison.decode!(body) |> Map.fetch!("error")
    {:error, error}
  end

  def parse({:error, error}), do: {:error, error}

  defp get_config(key, default \\ nil) when is_atom(key) do
    case Application.fetch_env(:stripy, key) do
      {:ok, {:system, env_var}} -> System.get_env(env_var)
      {:ok, value} -> value
      :error -> default
    end
  end

  defp get_config!(key) do
    get_config(key) || raise "stripy config #{key} is required"
  end
end
